azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,441 @@
+            self._trig_error(ValueError("Time is up for query timer"))
+
+        self._query_timer = Timer(constant.DEFAULT_TIMEOUT_INTERVAL, time_up_query)
+        self._query_timer.start()
+
+    def _wait_for_interval(self, event_data):
+        def time_up_polling():
+            self._polling_timer.cancel()
+            logger.info("Done waiting for polling interval of {} secs".format(polling_interval))
+            if result.operation_id is None:
+                self._trig_send_register_request(event_data)
+            else:
+                self._trig_poll(event_data)
+
+        result = event_data.args[0]
+        polling_interval = (
+            constant.DEFAULT_POLLING_INTERVAL
+            if result.retry_after is None
+            else int(result.retry_after, 10)
+        )
+
+        self._polling_timer = Timer(polling_interval, time_up_polling)
+        logger.info("Waiting for " + str(constant.DEFAULT_POLLING_INTERVAL) + " secs")
+        self._polling_timer.start()  # This is waiting for that polling interval
+
+    def _decode_complete_json_response(self, query_result, response):
+        """
+        Decodes the complete json response for details regarding the registration process.
+        :param query_result: The partially formed result.
+        :param response: The complete response from the service
+        """
+        decoded_result = json.loads(response)
+
+        decoded_state = (
+            None
+            if "registrationState" not in decoded_result
+            else decoded_result["registrationState"]
+        )
+        registration_state = None
+        if decoded_state is not None:
+            # Everything needs to be converted to string explicitly for python 2
+            # as everything is by default a unicode character
+            registration_state = RegistrationState(
+                None if "deviceId" not in decoded_state else str(decoded_state["deviceId"]),
+                None if "assignedHub" not in decoded_state else str(decoded_state["assignedHub"]),
+                None if "substatus" not in decoded_state else str(decoded_state["substatus"]),
+                None
+                if "createdDateTimeUtc" not in decoded_state
+                else str(decoded_state["createdDateTimeUtc"]),
+                None
+                if "lastUpdatedDateTimeUtc" not in decoded_state
+                else str(decoded_state["lastUpdatedDateTimeUtc"]),
+                None if "etag" not in decoded_state else str(decoded_state["etag"]),
+            )
+
+        registration_result = RegistrationResult(
+            rid=query_result.rid,
+            operation_id=query_result.operation_id,
+            status=query_result.status,
+            registration_state=registration_state,
+        )
+        return registration_result
+
+    def _decode_json_response(self, rid, retry_after, response):
+        """
+        Decodes the json response for operation id and status
```

Comment:
>Decodes the json response for operation id and status [](start = 8, length = 53)

This does not do what I'd expect from the function name

Reply:
the line on 413 does this right ? what kind of name are you thinking

---
In reply to: [282975735](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r282975735) [](ancestors = 282975735)